ggml-opencl.c

```diff
@@ -188,48 +186,143 @@ static cl_program build_program_from_source(cl_context ctx, cl_device_id dev, co
+
+    enum { NPLAT = 16, NDEV = 16 };
+
+    char text_buffer[1024] = {0};
+
+    platform = NULL;
+    device = NULL;
+
+    cl_platform_id platforms[NPLAT];
     cl_uint num_platforms;
-    clGetPlatformIDs(0, NULL, &num_platforms);
-    cl_platform_id* platforms = (cl_platform_id*)malloc(num_platforms*sizeof(cl_platform_id));
-    clGetPlatformIDs(num_platforms, platforms, NULL);
-    platform = platforms[plat_num];
-    char platform_buffer[1024];
-    clGetPlatformInfo(platform, CL_PLATFORM_NAME, sizeof(platform_buffer), &platform_buffer, NULL);
-    cl_uint num_devices;
-    clGetDeviceIDs(platform, CL_DEVICE_TYPE_ALL, 0, NULL, &num_devices);
-    cl_device_id* devices = (cl_device_id*)malloc(num_devices*sizeof(cl_device_id));
-    clGetDeviceIDs(platform, CL_DEVICE_TYPE_ALL, num_devices, devices, NULL);
-    device = devices[dev_num];
-    char device_buffer[1024];
-    clGetDeviceInfo(device, CL_DEVICE_NAME, sizeof(device_buffer), &device_buffer, NULL);
-    printf("Using Platform: %s Device: %s\n", platform_buffer, device_buffer);
-    context = clCreateContext(NULL, 1, &device, NULL, NULL, &err);
-    CL_CHECK(err, "clCreateContext");
-    queue = clCreateCommandQueue(context, device, CL_QUEUE_OUT_OF_ORDER_EXEC_MODE_ENABLE, &err);
-    CL_CHECK(err, "clCreateCommandQueue");
-
-    free(platforms);
-    free(devices);
-
-    program = build_program_from_source(context, device, clblast_dequant);
+    CL_CHECK(clGetPlatformIDs(NPLAT, platforms, &num_platforms));
+
+    char * GGML_OPENCL_PLATFORM = getenv("GGML_OPENCL_PLATFORM");
+    if (GGML_OPENCL_PLATFORM != NULL) {
+        unsigned plat_num;
+        if (sscanf(GGML_OPENCL_PLATFORM, " %u", &plat_num) == 1) {
+            if (plat_num >= num_platforms) {
+                fprintf(stderr, "ggml_opencl: There is no platform %d\n", plat_num);
+                exit(1);
+            } else {
+                platform = platforms[plat_num];
+                CL_CHECK(clGetPlatformInfo(platform, CL_PLATFORM_NAME, sizeof(text_buffer), &text_buffer, NULL));
+            }
+        } else {
+            for (unsigned i = 0; i < num_platforms; i++) {
+                CL_CHECK(clGetPlatformInfo(platforms[i], CL_PLATFORM_NAME, sizeof(text_buffer), &text_buffer, NULL));
+                if (strstr(text_buffer, GGML_OPENCL_PLATFORM) != NULL) {
+                    platform = platforms[i];
+                    break;
+                }
+            }
+        }
+        if (platform == NULL) {
+            fprintf(stderr, "ggml_opencl: no platform matching '%s' was found.\n", GGML_OPENCL_PLATFORM);
+            exit(1);
+        } else {
+            fprintf(stderr, "ggml_opencl: selecting platform: '%s'\n", text_buffer);
+        }
+    }
+
+    char * GGML_OPENCL_DEVICE = getenv("GGML_OPENCL_DEVICE");
+    if (GGML_OPENCL_DEVICE != NULL) {
+        cl_device_id devices[NDEV];
+        cl_uint num_devices;
+        CL_CHECK(clGetDeviceIDs(platform, CL_DEVICE_TYPE_ALL, NDEV, devices, &num_devices));
```

Comment:
`ggml_opencl: clGetDeviceIDs(platform, CL_DEVICE_TYPE_ALL, NDEV, devices, &num_devices) error -32 at ggml-opencl.c:233`

If I just select a device, it fails here with `CL_INVALID_PLATFORM`. I think it should select the first platform in this case.

Reply:
To select a device we need to know what platform it is in.

I have to think about this again. It shouldn't be this hard.

If we just put all the devices in all the platforms in a long list and the user selects from those, it would be easier to understand?

Reply:
Probably. But maybe we're also just overthinking it, this is a rare problem. Most users have one or two platforms with one or two devices at most.

Since many users will only have one platform, I think it would be enough to use the first platform if only a device is selected.